// react-highlight-within-textarea/lib/cjs/DecoratorFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecoratorFactory = void 0;
const react_1 = __importDefault(require("react"));
const draft_js_1 = require("draft-js");
const highlightToFlatStrategy_1 = require("./highlightToFlatStrategy");
const getMatches_1 = require("./getMatches");
const breakSpansByBlocks_1 = require("./breakSpansByBlocks");
class DecoratorFactory {
    constructor() {
        this.map = new Map();
        this.decoratorCache = new Map();
    }
    updateBlockSpans(blockSpans) {
        this.map.clear();
        for (const blockSpan of blockSpans) {
            const m = this.map.get(blockSpan.component) || new Map();
            this.map.set(blockSpan.component, m);
            const mm = m.get(blockSpan.className) || new Map();
            m.set(blockSpan.className, mm);
            const mmm = mm.get(blockSpan.blockKey) || new Map();
            mm.set(blockSpan.blockKey, mmm);
            mmm.set(blockSpan.spanStart - blockSpan.blockStart, blockSpan);
        }
    }
    componentFactory(component, className) {
        if (component == undefined) {
            return react_1.default.memo((props) => {
                return react_1.default.createElement("mark", { className: className }, props.children);
            }, (a, b) => a.decoratedText == b.decoratedText);
        }
        else {
            return (props) => {
                var _a, _b, _c;
                const block = (_c = (_b = (_a = this.map
                    .get(component)) === null || _a === void 0 ? void 0 : _a.get(className)) === null || _b === void 0 ? void 0 : _b.get(props.blockKey)) === null || _c === void 0 ? void 0 : _c.get(props.start);
                if (!block || !block.component) {
                    throw new Error("RHWTAInternalError");
                }
                return react_1.default.createElement(block.component, Object.assign({}, block), props.children);
            };
        }
    }
    strategyFactory(component, className) {
        return (block, callback) => {
            var _a, _b, _c;
            const blockSpans = ((_c = (_b = (_a = this.map
                .get(component)) === null || _a === void 0 ? void 0 : _a.get(className)) === null || _b === void 0 ? void 0 : _b.get(block.getKey())) === null || _c === void 0 ? void 0 : _c.values()) || [];
            for (const blockSpan of blockSpans) {
                const start = blockSpan.spanStart - blockSpan.blockStart;
                const end = blockSpan.spanEnd - blockSpan.blockStart;
                callback(start, end);
            }
        };
    }
    toDecorators(blockSpans) {
        var _a;
        const decorators = [];
        this.updateBlockSpans(blockSpans);
        for (const [component, m] of this.map.entries()) {
            for (const [className] of m.entries()) {
                let decorator = (_a = this.decoratorCache.get(component)) === null || _a === void 0 ? void 0 : _a.get(className);
                if (!decorator) {
                    decorator = {
                        component: this.componentFactory(component, className),
                        strategy: this.strategyFactory(component, className),
                    };
                    let c = this.decoratorCache.get(component) || new Map();
                    this.decoratorCache.set(component, c);
                    c.set(className, decorator);
                }
                decorators.push(decorator);
            }
        }
        return decorators;
    }
    create(contentState, highlight, text) {
        text = text || contentState.getPlainText();
        const sc = (0, highlightToFlatStrategy_1.highlightToFlatStrategy)(highlight);
        const matches = (0, getMatches_1.getMatches)(text, sc);
        const blockSpans = (0, breakSpansByBlocks_1.breakSpansByBlocks)(contentState, matches, text);
        const decorators = this.toDecorators(blockSpans);
        const draftDecorators = new draft_js_1.CompositeDecorator(decorators);
        return draftDecorators;
    }
}
exports.DecoratorFactory = DecoratorFactory;
